### PR TITLE
Test 0->0 correspondence for Unary Ops with Sparse CSR inputs

### PR DESCRIPTION
**Pull request by krshrimali**
Since there is no rule in PyTorch (Sparse CSR) for filling zeros, it was decided that only those ops will be supported which do not break 0->0 correspondence. To ensure that this rule is not broken, this PR aims to add a test to ensure this rule is not broken.

`sample_inputs_unary` may or may not generate a zero in the sample input. Hence, this separate test is good for validating the rule, and the support for Sparse CSR.

cc @nikitaved @pearu @cpuhrsch

### CODE REVIEW

**Review by IvanYashchuk — APPROVED** [review]

File: test/test_sparse_csr.py [hunk 1]
```diff
@@ -1185,6 +1185,25 @@ def test_sparse_csr_unary_consistency(self, device, dtype, op):
...
+        zero = torch.zeros((1, 2), dtype=dtype, device=device)
+
+        output = op(zero)
+        expected = zero.to(output.dtype)
+        self.assertEqual(output, expected, f"This operator ({op.name}) should not be supported for "
+                         + "Sparse CSR as it breaks 0->0 correspondence.")
+        zero = zero.to_sparse_csr()
+        output = op(zero).to_dense()
+        expected = zero.to_dense().to(output.dtype)
+        self.assertEqual(output, expected, f"0->0 correspondence is not satisfied for {op.name}.")
```
Comment (pearu) [inline test/test_sparse_csr.py:1205]: Notice that `zero.to_sparse_csr().values().numel() == 0`. Hence, also `op(zero.to_sparse_csr()).values().numel() == 0` must hold. In fact, this condition also tests that the unary `op` on sparse CSR input will not produce CSR tensors with values containing explicit zeros. So:
```suggestion
        self.assertEqual(op(zero.to_sparse_csr()).values().numel(), 0,
                         f"0->0 correspondence is not satisfied for {op.name} or explicit zeros in CSR result values.")
```

**Review by pearu — CHANGES_REQUESTED** [review]
I have a nit to make testing `op(csr_zeros) -> csr_zeros` stronger in the sense that `op` should not produce from zeros input a CSR tensor with values containing explicit zeros.

On the other hand, if the CSR input to `op` contains explicit zeros, the op should produce a tensor with the corresponding explicit zeros. So, could you also add a test for the following input:
```
t = torch.sparse_csr_tensor([0, 1], [1], [0])
```
? Here you can use the `op(t).to_dense() == t.to_dense()` approach in addition to the `op(t).values().numel() == 1` check.

**Comment by IvanYashchuk** [comment]
> I have a nit to make testing op(csr_zeros) -> csr_zeros stronger in the sense that op should not produce from zeros input a CSR tensor with values containing explicit zeros.

Well, it should. In many operations, we shouldn't modify the sparsity pattern. Requiring compressing zeros after any operation would unnecessarily complicate the implementation and would result in performance regression as modifying the indices of CSR matrix is an expensive operation.

> if the CSR input to op contains explicit zeros, the op should produce a tensor with the corresponding explicit zeros

This is what we have implemented now, and it's a cheaper alternative to the first option.

**Comment by pearu** [comment]
> Well, it should.

This is wrong. I think my message was not clear enough then. The unary `ops` operate on values only. My statement is that a unary operation should not introduce new items to values such as explicit zeros, although, mathematically one would have correct results after densification. For example (using COO for simplicity):
1. `sin(sparse_coo_tensor([[0], [0]], [pi/2], size=(1, 2))) -> sparse_coo_tensor([[0, 0], [0, 1]], [1, 0]])` is not allowed as the result values size has increased (it has introduced an explicit output zero corresponding to implicit input zero).
1. `sin(sparse_coo_tensor([[0, 0], [0, 1]], [pi/2, 0], size=(1, 2))) -> sparse_coo_tensor([[0, 0], [0, 1]], [1, 0]])` is allowed (explicit output zero corresponds to explicit input zero).

> In many operations, we shouldn't modify the sparsity pattern.

Exactly.

The current test does not catch this, hence my suggestion above.

**Comment by krshrimali** [comment]
Hi @pearu - thank you for your review and suggestion. I've incorporated them in this PR, could you please take a look again whenever you find the time?

File: test/test_sparse_csr.py [hunk 2]
```diff
@@ -1277,6 +1277,33 @@ def test_sparse_csr_consistency(self, device, dtype, op):
...
+
+        for (output, expected) in [
+                (output_zero, expected_zero),
+                (output_explicit_zeros, expected_explicit_zeros)
+        ]:
+            self.assertEqual(output, expected, f"This operator ({op.name}) should not be supported for "
+                             + "Sparse CSR as it breaks 0->0 correspondence.")
+
+        for (inp, expected_zero_count) in [(zero.to_sparse_csr(), 0), (tensor_explicit_zeros, 1)]:
+            self.assertEqual(op(inp).values().numel(), expected_zero_count,
```
Comment (pearu) [inline test/test_sparse_csr.py:1304]: To test that op does not change the sparsity pattern, I suggest using
```suggestion
            self.assertEqual(op(inp).values().numel(), inp.values().numel(),
```
and eliminate `expected_zero_count`.

File: test/test_sparse_csr.py [hunk 3]
```diff
@@ -1277,6 +1277,33 @@ def test_sparse_csr_consistency(self, device, dtype, op):
...
+                             f"0->0 correspondence is not satisfied for {op.name} or explicit zeros in CSR result values.")
```
Comment (pearu) [inline test/test_sparse_csr.py:1305]: Not sure if this message corresponds to what is tested here. So
```suggestion
                             f"{op.name} fails to preserve sparsity pattern.")
```

File: test/test_sparse_csr.py [hunk 4]
```diff
@@ -1277,6 +1277,33 @@ def test_sparse_csr_consistency(self, device, dtype, op):
...
+
+        output_explicit_zeros = op(tensor_explicit_zeros).to_dense()
+        expected_explicit_zeros = tensor_explicit_zeros.to_dense().to(output_explicit_zeros.dtype)
+
+        for (output, expected) in [
+                (output_zero, expected_zero),
+                (output_explicit_zeros, expected_explicit_zeros)
+        ]:
+            self.assertEqual(output, expected, f"This operator ({op.name}) should not be supported for "
+                             + "Sparse CSR as it breaks 0->0 correspondence.")
```
Comment (pearu) [inline test/test_sparse_csr.py:1301]: Nit: using `+` is unnecessary.

**Review by pearu — CHANGES_REQUESTED** [review]
I have a couple of nits but overall looks good.

File: test/test_sparse_csr.py (hunk 2 above)
Comment (krshrimali) [reply test/test_sparse_csr.py]: Got it, thanks Pearu. Should be fixed now.

File: test/test_sparse_csr.py (hunk 4 above)
Comment (krshrimali) [reply test/test_sparse_csr.py]: Ah, right. Should be fixed now :)

**Review by pearu — APPROVED** [review]
LGTM! Thanks, @krshrimali !

**Review by cpuhrsch — APPROVED** [review]

**Pull request closed without merging**